### DISCUSSION TITLE
Ti.UDP update (please help!)

### DISCUSSION THREAD

**Discussion by focussing** [appcelerator/titanium_mobile#13221] [Q&A]
Hi everybody, could anyone please help me to update the very old Ti.UDP module. This module was originally created by Appcelerator but no longer updated.
I need to update an Android app built back in 2019 using Ti SDK6.3.0, but I cannot get it to work anymore:

`TiExceptionHandler: (main) [0,223] - Message: Uncaught ReferenceError: global is not defined`

I tried lots of things, node version, other SDK’s, etc. And I cannot build the module to get it run at SDK 8.2 or so.

```
[INFO] :   Found Titanium module id=ti.udp version=3.1.0 platform=android deploy-type=test path=/Users/raymond/Library/Application Support/Titanium/modules/android/ti.udp/3.1.0
[WARN] :   The module ti.udp does not support the ABI: arm64-v8a
[WARN] :   It only supports the following ABIs: armeabi-v7a, x86
[WARN] :   Your application will most likely encounter issues
```

I use the same module for the iOS variant of the app and that works super with the latest  SDK 10.1.1 in combination with Xcode version 13.

The best way of course would be to update the Android app and then use a modernized Ti.UDP module...

The module can be downloaded from here: www.focussing.nl/ti.udp-master.zip 

Any help is very much appreciated!

Best regards,
Raymond

**Comment by m1ga** [comment]
Is that the same version as https://github.com/appcelerator-archive/ti.udp ?
Its just a rebuild with min SDK 9:
[ti.udp-android-4.0.0.zip](https://github.com/appcelerator/titanium_mobile/files/7758224/ti.udp-android-4.0.0.zip)

But are you sure that the error is by that module? Try that file and if it works I'll make a PR to the repo so its linked

**Comment by focussing** [comment]
Thanks a lot! You are great!!

Until now I used the version at www.focussing.nl/ti.udp-master.zip <http://www.focussing.nl/ti.udp-master.zip> (Ti.UDP version 3.1.0) which originally was derived from https://github.com/appcelerator-archive/ti.udp <https://github.com/appcelerator-archive/ti.udp>.

That version 3.1.0 works fine in my iOS app which was built using Ti SDK 10.1.1 and released in the AppStore.
It also worked fine in Android app based on Ti SDK 6.3.0 back in 2019.


I tested your Ti.UPD 4.0.0 using Ti SDK 10.1.1.GA using an android device.

It builds perfectly, but I get a runtime exception:

[INFO]  {"host":"192.168.1.201","data":"STA"}
[ERROR] TiExceptionHandler: (main) [100,4959] /app.js:98
[ERROR] TiExceptionHandler: Alloy.Globals.sockets.socket.sendString({
[ERROR] TiExceptionHandler:                              ^
[ERROR] TiExceptionHandler: Error: Unknown Java Exception occurred
[ERROR] TiExceptionHandler:     at Object.Alloy.Globals.sendString (/app.js:98:30)
[ERROR] TiExceptionHandler:     at Titanium.<anonymous> (/alloy/controllers/index.js:201:15)
[ERROR] TiExceptionHandler:
[ERROR] TiExceptionHandler:     android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1273)
[ERROR] TiExceptionHandler:     libcore.io.BlockGuardOs.sendto(BlockGuardOs.java:276)
[ERROR] TiExceptionHandler:     libcore.io.IoBridge.sendto(IoBridge.java:509)
[ERROR] TiExceptionHandler:     java.net.PlainDatagramSocketImpl.send(PlainDatagramSocketImpl.java:184)
[ERROR] TiExceptionHandler:     java.net.DatagramSocket.send(DatagramSocket.java:305)
[ERROR] TiExceptionHandler:     ti.udp.SocketProxy.sendString(SocketProxy.java:202)
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.runtime.v8.V8Function.nativeInvoke(Native Method)
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.runtime.v8.V8Function.callSync(V8Function.java:55)
[ERROR] TiExceptionHandler:     org.appcelerator.kroll.runtime.v8.V8Function.call(V8Function.java:41)
[ERROR] TiExceptionHandler:     ti.modules.titanium.TitaniumModule$Timer.run(TitaniumModule.java:145)
[ERROR] V8Exception: Exception occurred at /app.js:98: Uncaught Error: Unknown Java Exception occurred

In the iOS situation using the same code the Ti.UDP works perfectly.

I hope this can be fixed; that would be the perfect Xmas present!!


Best regards, 
Raymond

**Reply by m1ga** [reply]
My diff tool says that both projects (your zip and ti.udp) are identical. Did you made changes in a different branch? The zip is called `-master` so perhaps its the wrong branch.

About that error:
it should catch an error at that point (IOException) but I've changed it to a genenarl `Exception` in this version:
[ti.udp-android-4.0.0.zip](https://github.com/appcelerator/titanium_mobile/files/7761515/ti.udp-android-4.0.0.zip)
So please try this and check out `error` event. Since it looks like it is going into the sendString method and forwarding it to internal methods.

Searching around leads to issues like: https://stackoverflow.com/questions/25020660/datagram-socket-udp-java-android where they say it should run in a different thread which I've added here:
[ti.udp-android-4.0.1.zip](https://github.com/appcelerator/titanium_mobile/files/7761579/ti.udp-android-4.0.1.zip)
but I don't have anyway to test it so its just a blind compile :)

**Comment by focussing** [comment]
Thank you very much for the fast reply.
I tested both versions, and there is no error generated anymore :)
The socket is started but when I send messages to the electronics board nothing happens, I don't see anything receiving on the board and no reply is sent from the board.
The console:

-- Start application log -----------------------------------------------------
[INFO]  art: Late-enabling -Xcheck:jni
[INFO]  art: VMHOOK: rlim_cur : 0 pid:21456
[INFO]  TiApplication: (main) [1,1] checkpoint, app created.
[INFO]  TiApplication: (main) [346,347] Analytics have been disabled
[INFO]  TiApplication: (main) [0,347] Titanium Javascript runtime: v8
[INFO]  TiRootActivity: (main) [0,0] checkpoint, on root activity create, savedInstanceState: null
[WARN]  art: Before Android 4.1, method android.graphics.PorterDuffColorFilter androidx.vectordrawable.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
[INFO]  TiRootActivity: (main) [0,0] checkpoint, on root activity resume. activity = nl.focussing.bosan.BosanActivity@1c76ed2
[INFO]  testproc 1.2.2 (Powered by Titanium 10.1.1.2448cfa671)
**[WARN]  SocketProxy: Setter method deprecated, please use "obj.bufferSize = val;" or "obj['bufferSize'] = val;" instead.
[INFO]  SocketProxy: (main) [221,221] Socket Started!**
[INFO]  I/Adreno-EGL: <qeglDrvAPI_eglInitialize:379>: EGL 1.4 QUALCOMM build: Nondeterministic_AU_msm8974_LA.BF.1.1.3__release_AU (I3fa967cfef)
[INFO]  I/Adreno-EGL: OpenGL ES Shader Compiler Version: E031.28.00.02
[INFO]  I/Adreno-EGL: Build Date: 09/29/15 Tue
[INFO]  I/Adreno-EGL: Local Branch: mybranch14683032
[INFO]  I/Adreno-EGL: Remote Branch: quic/master
[INFO]  I/Adreno-EGL: Local Patches: NONE
[INFO]  I/Adreno-EGL: Reconstruct Branch: NOTHING
[INFO]  WIFI connected
[INFO]  scanModules
[INFO]  Alloy.Globals.nrOfModules: 1
[INFO]  Project built successfully in 55s 917ms
**[INFO]  {"host":"192.168.1.201","data":"STA"}
[INFO]  {"host":"192.168.1.201","data":"STA"}**

**Reply by m1ga** [reply]
I don't see a `Data Sent!` in there so I guess it stopped before.
Please try this one:
[ti.udp-android-4.0.2.zip](https://github.com/appcelerator/titanium_mobile/files/7761772/ti.udp-android-4.0.2.zip)
I've added a log output for the error and a "trying to send..." log before the function that made an error before.

Also:
use `socket.bufferSize = ...` now with this version. Getters/setters are removed in 10.x and I've updated the module now so it won't throw that warning.

**Comment by focussing** [comment]
The buffersize warning has gone now.
This is the log:

[INFO]  WIFI connected
[INFO]  scanModules
[INFO]  Alloy.Globals.nrOfModules: 1
[INFO]  {"host":"192.168.1.201","data":"STA"}
[ERROR] SocketProxy: (main) [3368,3690] ERROR: android.os.NetworkOnMainThreadException
[INFO]  {"host":"192.168.1.201","data":"STA"}
[ERROR] SocketProxy: (main) [1977,5667] ERROR: android.os.NetworkOnMainThreadException

**Reply by m1ga** [reply]
oh sorry, I've add my changes to `sendBytes` instead of `sendString`! One sec

Try this please: [ti.udp-android-4.0.3.zip](https://github.com/appcelerator/titanium_mobile/files/7761908/ti.udp-android-4.0.3.zip) @focussing

**Comment by focussing** [comment]
YOU ARE AMAZING

It works like a charm!!
Thank you so much!

**Reply by m1ga** [reply]
Awesome :+1: I've added everything to this PR: https://github.com/appcelerator-archive/ti.udp/pull/11
I've removed my custom logs and put the other logs into `--log-level debug` so it won't be in the normal logs.

**Comment by focussing** [comment]
There is some weird thing going on, it only works one time. When I close the app and start it again it does not work anymore

**Reply by focussing** [reply]
If I build it again it is installed on the devices and starts up it works. If I close the app it does not work anymore.

**Reply by m1ga** [reply]
app just going to background or closing (`exitOnClose:true`) e.g. do you see the splashscreen again when you open it the 2nd time? Is there an error log the second time?

**Reply by focussing** [reply]
There is no exitOnClose:true in my code.
I close all running apps and when I restart it the splashscreen comes up, but without logging (vscode does not show it anymore). 
The app starts up but with no communication over UDP/IP so it does not find the board. It does not crash or so.
Only when I "build for devices" in vscode it starts working again with communicating with the board.
Could it be related to this extra thread you added?

**Reply by m1ga** [reply]
you can always use `adb logcat | grep socketProxy` to see the logs without VSCode. If it doesn't connect to the board the error should be in the `start` method. But check the adb log, maybe there is something helpful

**Reply by focussing** [reply]
how do I do that? 
macbook:bosan raymond$ adb logcat | grep socketProxy
-bash: adb: command not found

**Comment by focussing** [comment]
sorry;

macbook:bosan raymond$ /Users/raymond/android-sdk/platform-tools/adb logcat | grep socketProxy 

but then it waits... nothing show when I start the app

**Reply by m1ga** [reply]
can you try `/Users/raymond/android-sdk/platform-tools/adb logcat | grep -i socketProxy`

**Reply by focussing** [reply]
same...
I give that command in terminal, and next I start the app on my android. that's right?
vscode is still running but does not show logging. It always stops when the app is closed.

**Reply by m1ga** [reply]
yes starting it in a terminal is correct. You can also just use `/Users/raymond/android-sdk/platform-tools/adb logcat` but that will show many logs.

**Reply by focussing** [reply]
/Users/raymond/android-sdk/platform-tools/adb logcat | grep -i socketProxy does show info when build/run from vscode. it stops showing when the app is closed. so basically shows the same as VScode

**Reply by m1ga** [reply]
@focussing still on slack? I've send you a ping there. Perhaps its quicker to debug it there :)

**Comment by m1ga** [comment]
to finish this up here too:
I've updated the module at https://github.com/appcelerator-archive/ti.udp/pull/11
and it works again now :+1:
